docs/proposals/14-reversed-cluster-vpn.md

```diff
@@ -0,0 +1,276 @@
+* We do not intend to provide continuous availability to the shoot-seed VPN connection. We expect the availability to be comparable to the existing solution.
+
+## Proposal
+
+The propsal is depicted in the following diagram:
+
+![alt text](assets/ReversedTunnelVPN.png "Reversed Tunnel VPN")
+
+We have added an OpenVPN server pod (`vpn-seed-server`) to each control plane. This OpenVPN server is configured with the OpenVPN `client-to-client` option that allows clients to talk to each other. The OpenVPN client in the shoot cluster (`vpn-shoot-client`) as well as the Open VPN client sidecars in the seed (`vpn-seed-client`) connect to the OpenVPN server.
+
+The three containers `vpn-seed-server`, `vpn-seed-client`, and `vpn-shoot-client` are new containers and are not related to containers in the github.com/gardener/vpn project. We will create a new project github.com/gardener/vpn2 for these containers. With this solution we intend to supersede the containers from the github.com/gardener/vpn project.
+
+A service `openvpn-server` is created for each control plane in its namespace. The `vpn-seed-client` sidecars connect directly to this service.
+
+The `vpn-shoot-client` pod connects to the correct `vpn-seed-server` service via the SNI passthrough proxy introduced with [SNI Passthrough proxy for kube-apiservers](https://github.com/gardener/gardener/blob/master/docs/proposals/08-shoot-apiserver-via-sni.md) port 443. No additional port needs to be opened. We re-use the SNI infrastructure in order to avoid the need to have one external load balancer per shoot cluster in the seed.
+
+Shoot OpenVPN clients connect to the correct OpenVPN Server using the http proxy feature provided by OpenVPN. A configuration is added to the envoy proxy to detect http proxy requests and forward the connection attempt to the correct OpenVPN server.
+
+The restriction regarding the `192.268.123.0/24` network range in the current VPN solution still applies to this proposal. No other restrictions are introduced.
+
+### Performance and Scalability
+
+We do expect performance and throughput to be slightly lower compared to the existing solution. This is because the OpenVPN server acts as an additional hop and must decrypt and re-encrypt traffic that passes through. As there are no low latency, or high thoughput requirements for this connection we do not assume this to be an issue.
+
+### Availablity and Failure Scenarios
+
+This solution re-uses multiple instances of the envoy component used for the kube-apiserver endpoints. We assume that the availability for kube-apiservers is good enough for the cluster VPN as well. 
+
+The OpenVPN client- and server pods are singleton pods in this approach and therefore are affected by potential failures and during cluster-, and control plane updates. Potentiall outages are only restricted to single shoot clusters and are comparable to the situation with the existing solution today.
```

Comment:
```suggestion
The OpenVPN client and server pods are singleton pods in this approach and therefore are affected by potential failures and during cluster-, and control plane updates. Potential outages are only restricted to single shoot clusters and are comparable to the situation with the existing solution today.
```

Reply:
Hm, OK, it's not getting "worse", but is running them with multiple replicas even possible? If we now invest into a new "VPN solution" then we probably want to tackle such technical debt?

Reply:
I do not think that it can be solved just using replicas. However, I think that it should be possible to add a second/third/etc. pair of `vpn-seed-server`/`vpn-shoot-client`. This would allow the `vpn-seed-client` (in the kube-apiserver pod) to use a fallback if necessary, i.e. connect to one first and then switch to the second/third if the first goes down.
However, this adds considerable complexity and increases the footprint. I am also not sure if it is justified in terms of potentially gained availability. Therefore, I would propose to start with the simply solution and only switch to a more complex solution if needed.

Reply:
> , I would propose to start with the simply solution and only switch to a more complex solution if needed.

Yes, fair enough, I was more interested whether this is possible in the first place or whether this is a limitation. 

> However, this adds considerable complexity and increases the footprint.

Which kind of complexity is added? Can you talk about this in the GEP?

Reply:
If the availability issue could be solved with simple replicas, I would not talk about additional complexity. The additional complexity here would be to have kind of duplicate deployments and the `vpn-seed-client` falling back/switching between them.
A section can be added to the GEP for sure.

Reply:
Not sure what you mean with "to have kind of duplicate deployments" (this is just about having multiple replicas, doesn't this work with OpenVPN in general or why do you mention this when we talk about "complexity)?

As for "the vpn-seed-client falling back/switching between them.", is there anything special we need to pay attention to? Why doesn't it work out of the box (if the seed-client already established a connection then it stays with the respective replica of the server, but if it needs to establish a new connection then the Kubernetes service discovery will randomly select one of the many available servers)?

Reply:
After internal alignment with @marwinski and @DockToFuture, our proposal is to start with the current proposal as is, i.e. with one instance of the `vpn-seed-server` and one instance of the `vpn-shoot-client` (+ as many `vpn-seed-client` instances as there are kube-apiservers). Then we should check our monitoring data concerning the real availability to understand whether we face an issue or not. In case we would see enough evidence that there is an issue with the approach, we could extend it. However, at the moment we see this as a high investment/low benefit task, i.e. not worthwhile.

From the `vpn-seed-client`'s perspective, it might look like there are only replicas required and a kubernetes service might be sufficient for the routing from `vpn-seed-client` to `vpn-seed-server`. Unfortunately, it is not as simple as this. It is necessary that each `vpn-seed-server` instance has a `vpn-shoot-client` connected to it. Otherwise, it would be rather useless for the `vpn-seed-client` as no connectivity to the shoot would exist. This means that you need to scale not only the `vpn-seed-server`, but also the `vpn-shoot-client` and ensure that each `vpn-seed-server` is connected to exactly one of the `vpn-shoot-client` instances. As far as I know, simple service routing with replicas would not ensure that we have a one-to-one mapping here. Feel free to convince me otherwise.
Hence, we could for example create `vpn-seed-server-1` and `vpn-seed-server-2` on the seed and `vpn-shoot-client-1` and `vpn-shoot-client-2` on the shoot. Then we would ensure that `vpn-shoot-client-1` connects to `vpn-seed-server-1` and `vpn-shoot-client-2` connects to `vpn-seed-server-2`. For this, explicit endpoints for `vpn-seed-server-1` and `vpn-seed-server-2` are required. The `vpn-seed-client` can connect to any of the `vpn-seed-server`. Therefore, in this case a kubernetes service with session stickiness might be used. (OpenVPN also allows a client to configure several servers and fall back if one does not work.)

I hope this clarifies the topic.